src/services/dynamicsDal.service.js

```diff
@@ -18,99 +18,103 @@ module.exports = class DynamicsDalService {
+  async create (query, dataObject) {
+    this._setCallSpecificRequestOptions(query, 'POST', dataObject)
+    const result = await this._call(query, dataObject)
+    const id = this._extractId(result)
+    return id
   }
 
-  update (dataObject, query) {
-    return this._commit(query, 'PATCH', dataObject)
+  async update (query, dataObject) {
+    this._setCallSpecificRequestOptions(query, 'PATCH', dataObject)
+    this._call(query, dataObject)
   }
 
-  search (query) {
-    return this._query(query)
+  async search (query) {
+    this._setCallSpecificRequestOptions(query, 'GET')
+    const result = await this._call(query)
+    return result
   }
 
-  _query (query) {
-    return new Promise((resolve, reject) => {
-      // Combine the path to Dynamics and the query and add it to our request options
-      this.crmRequestOptions.path = `${config.dynamicsWebApiPath}${query}`
+  _setCallSpecificRequestOptions (query, method, dataObject = undefined) {
+    // Combine the path to Dynamics and the query and add it to our request options
+    this.crmRequestOptions.path = `${config.dynamicsWebApiPath}${query}`
+    this.crmRequestOptions.method = method
 
-      this.crmRequestOptions.method = 'GET'
+    // Set the content length
+    if (dataObject) {
+      const contentLength = Buffer.byteLength(JSON.stringify(dataObject))
+      this.crmRequestOptions.headers['Content-Length'] = contentLength
+    }
+  }
 
+  _call (query, dataObject = undefined) {
+    return new Promise((resolve, reject) => {
       // Query Dynamics for the data via a HTTPS request
       const crmRequest = https.request(this.crmRequestOptions, function (response) {
+        response.setEncoding('utf8')
+
         // We use an array to hold the response parts in the event we get multiple parts returned
         const responseParts = []
-        response.setEncoding('utf8')
         response.on('data', function (chunk) {
           responseParts.push(chunk)
         })
         response.on('end', function () {
           console.log(`Dynamics query response: Status Code: ${response.statusCode} Message: ${response.statusMessage}`)
 
-          if (response.statusCode === 200) {
-            // Parse the response JSON
-            resolve(JSON.parse(responseParts.join('')))
-          } else {
-            console.error(responseParts.join(''))
-            const message = `Unknown response from Dynamics. Code: ${response.statusCode} Message: ${response.statusMessage}`
-            reject(message)
+          switch (response.statusCode) {
+            case 200:
+              // Parse the response JSON
+              resolve(JSON.parse(responseParts.join('')))
+              break
+            case 204:
+              resolve(response.headers['odata-entityid'])
+              break
+            default:
+              const message = `Unknown response from Dynamics. Code: ${response.statusCode} Message: ${response.statusMessage}`
```

Comment:
This duplicates the message on line 63 – for example, if I remove the Dynamics password and trigger a 401, I get:

```
Dynamics query response: Status Code: 401 Message: Unauthorized
Unable to get Dynamics solution details: Unknown response from Dynamics. Code: 401 Message: Unauthorized
Unknown response from Dynamics. Code: 401 Message: Unauthorized
```

Might be overkill? I think that middle line is coming from the dynamicsSolution model.